Migrate away from docker.io

edsantiago opened the pull request:
CI and system tests currently pull some images from docker.io.
Eliminate that, by:

  - building a custom image containing much of what we need
    for testing; and
  - copying other needed images to quay.io

The principal change is to create a new quay.io/libpod/testimage,
using the new test/system/build-testimage script, instead of
relying on quay.io/libpod/alpine_labels. We also switch to
using a hardcoded :YYYYMMDD tag, instead of :latest, in an
attempt to futureproof our CI. This image includes 'httpd'
from busybox-extras, which we use in our networking test
(previously we had to pull and run busybox from docker.io).

The testimage can and should be extended as needed for future
tests, e.g. adding test file content or other useful tools.

For the '--pull' tests which require actually pulling from
the registry, I've created an image with the same name but
tagged :00000000 so it will never be pulled by default.
Since this image is only used minimally, it's just busybox.

Unfortunately there remain two cases we cannot solve in
this tiny alpine-based image:

  1) docker registry
  2) systemd

For those, I've (manually) run:

    podman pull [ docker.io/library/registry:2.7 | registry.fedoraproject.org/fedora:31 ]
    podman tag !$ quay.io/...
    podman push !$

...and amended the calling tests accordingly.

I've tried to make the the smallest reasonable diff, not the
smallest possible one. I hope it's a reasonable tradeoff.

Signed-off-by: Ed Santiago <santiago@redhat.com>

edsantiago commented:
Marked `WIP` because I'd like feedback on the image name: `testimage` seems too undescriptive.

Until naming is resolved, images live in my personal `quay.io` namespace instead of `quay.io/libpod`.

edsantiago commented on test/system/500-networking.bats:
```diff
@@ -12,7 +12,7 @@ load helpers
 
     HOST_PORT=8080
-    SERVER=http://localhost:$HOST_PORT
+    SERVER=http://127.0.0.1:$HOST_PORT
```
This is completely weird. When running podman as root, 'localhost' no longer seems to work, at least with alpine: `wget` barfs with ECONNREFUSED. It works fine with rootless. I gave up trying to track it down, and explicit IPv4 home seems like a better idea anyway.

rhatdan replied:
Weird, but localhost is hard coded within container/image and podman to represent locally built images.

edsantiago commented on test/system/helpers.bash:
```diff
@@ -5,9 +5,9 @@ PODMAN=${PODMAN:-podman}
-PODMAN_TEST_IMAGE_NAME=${PODMAN_TEST_IMAGE_NAME:-"alpine_labels"}
-PODMAN_TEST_IMAGE_TAG=${PODMAN_TEST_IMAGE_TAG:-"latest"}
+PODMAN_TEST_IMAGE_USER=${PODMAN_TEST_IMAGE_USER:-"edsantiago"}
+PODMAN_TEST_IMAGE_NAME=${PODMAN_TEST_IMAGE_NAME:-"testimage"}
```
Please offer suggestions here.

rhatdan replied:
I am fine with testimage?

edsantiago replied:
Okay, going once...

rhatdan commented on test/system/260-sdnotify.bats:
```diff
@@ -116,8 +116,7 @@ function _assert_mainpid_is_conmon() {
     # FIXME: is there a smaller image we could use?
-    _FEDORA=registry.fedoraproject.org/fedora:31
-
+    local _FEDORA="$PODMAN_TEST_IMAGE_REGISTRY/$PODMAN_TEST_IMAGE_USER/fedora:31"
```
Should we update this to 32?  Or eliminate the version altogether?

rhatdan commented on test/system/build-testimage:
```diff
@@ -0,0 +1,59 @@
+#    busybox-extras  : provides httpd needed in 500-networking.bats
+#
+podman rmi -f testimage &> /dev/null || true
+podman build --squash-all -t testimage - <<EOF
```
Would we want to switch this to --timestamp=0 or some other timestamp, to keep it consistent once this is available to Podman?

edsantiago replied:
I considered that but couldn't think of a reason - it'll be useful to have a number of tools built-in, and it would be **super** useful to figure out a way some day to include `registry` and systemd, but if/when that happens we'll change the `YMD` tag. Do you see value in reproducible/comparable images?

rhatdan reviewed (approved):
A few questions, and need to get rid of the edsantiago references 
but overall LGTM

vrothberg commented:
@edsantiago, what's the motivation to migrate away? I assume the latest policy changes of Docker Hub but it would be nice to mention it in the commit message for historical reasons.

vrothberg reviewed (commented):
Other than Dan's comments, LGTM

edsantiago commented:
>what's the motivation to migrate away? I assume the latest policy changes of Docker Hub but it would be nice to mention it in the commit message for historical reasons.

Yes, Docker Hub pull limit. I updated commit message to reflect that. 

Since `testimage` seems OK, I've pushed those images to `quay.io/libpod`, made them public, updated the reference in `helpers.bash`, and pushed.

rhatdan commented:
/lgtm
